lib/ui/painting.dart

```diff
@@ -2869,44 +2906,79 @@ class ImageFilter {
 
   final Float64List _data;
   final FilterQuality? _filterQuality;
+  final ColorFilter? _colorFilter;
+  final ImageFilter? _innerFilter;
+  final ImageFilter? _outerFilter;
   final int _type;
-  _ImageFilter? _nativeFilter;
+  late final _ImageFilter _nativeFilter = _makeNativeImageFilter();
 
   // The type of SkImageFilter class to create for Skia.
-  static const int _kTypeBlur = 0;   // MakeBlurFilter
-  static const int _kTypeMatrix = 1; // MakeMatrixFilterRowMajor255
+  static const int _kTypeComposed = 0;    // SkImageFilters::Compose
+  static const int _kTypeBlur = 1;        // MakeBlurFilter
+  static const int _kTypeMatrix = 2;      // MakeMatrixFilterRowMajor255
+  static const int _kTypeColorFilter = 3; // SkImageFilters::ColorFilter
 
   @override
   bool operator ==(Object other) {
     return other is ImageFilter
         && other._type == _type
         && _listEquals<double>(other._data, _data)
-        && other._filterQuality == _filterQuality;
+        && other._filterQuality == _filterQuality
+        && other._outerFilter == _outerFilter
+        && other._innerFilter == _innerFilter
+        && other._colorFilter == _colorFilter;
   }
 
-  _ImageFilter _toNativeImageFilter() => _nativeFilter ??= _makeNativeImageFilter();
+  _ImageFilter _toNativeImageFilter() => _nativeFilter;
 
   _ImageFilter _makeNativeImageFilter() {
     switch (_type) {
+      case _kTypeComposed:
+        return _ImageFilter.composed(this);
       case _kTypeBlur:
         return _ImageFilter.blur(this);
       case _kTypeMatrix:
         return _ImageFilter.matrix(this);
+      case _kTypeColorFilter:
+        return _ImageFilter.fromColorFilter(this);
       default:
         throw StateError('Unknown mode $_type for ImageFilter.');
     }
   }
 
   @override
-  int get hashCode => hashValues(_filterQuality, hashList(_data), _type);
+  int get hashCode => hashValues(_innerFilter, _outerFilter, _filterQuality, hashList(_data), _type, _colorFilter);
+
+  String? get _shortDescription {
+    switch (_type) {
+      case _kTypeComposed:
+        final String? outer = _outerFilter?._shortDescription;
+        final String? inner = _innerFilter?._shortDescription;
+        return (inner != null && outer != null) ? '$inner -> $outer' : null;
```

Comment:
I'm not sure that stringifying before testing matters. What was the intent there? Also, how do the inner/outer end up null, doesn't the constructor enforce non-null?

The else should return the one that wasn't null, though. I think you need at least 2 nested `?:` as in:

`outer == null ? inner : (inner == null ? outer : '$inner -> $outer');`

Reply:
`_innerFilter` and `_outerFilter` are guaranteed to be non-null in runtime. `_shortDescription` has the type `String?` because I'm using null to propagate errors. It returns null when any of the filters being composed has a bad `_type`.

Reply:
I didn't want `toString()` to return something like "ImageFilter.compose(Unknown ImageFilter type. This is an error. If you\'re seeing this, please file an issue at https://github.com/flutter/flutter/issues/new. -> blur -> blur)". But I guess it's better to tell the developer which filter went wrong precisely. Updating.

Reply:
Now `_shortDescription` returns a `String` instead of `String?`. The final `toString` method will return something like "ImageFilter.compose(unknown(type: 9) -> blur -> blur)" in case of error.